packages/renderer-commonmark/src/index.ts

```diff
@@ -103,114 +164,191 @@ export default class CommonmarkRenderer extends Renderer {
+   * A `code` span can be inline or as a block:
+   *
+   * ```js
+   * function () {}
+   * ```
    */
-  *'link'(props: { href: string }): IterableIterator<string> {
-    let [before, text, after] = yield* split();
-    return `${before}[${text}](${props.href})${after}`;
+  *'code'(props: { style: CodeStyle, info?: string }, state: State): IterableIterator<string> {
+    state.push({ isPreformatted: true, htmlSafe: false });
+    let text = yield;
+    state.pop();
+    let code = text.join('');
+
+    if (props.style === 'fence') {
+      code = '\n' + code;
+      let info = props.info || '';
+      let newlines = '\n';
+      if (state.get('isList') && state.get('nextAnnotation')) {
+        newlines += '\n';
+      }
+
+      if (code.indexOf('```') !== -1) {
+        return `~~~${info}${code}~~~${newlines}`;
+      } else {
+        return `\`\`\`${info}${code}\`\`\`${newlines}`;
+      }
+    } else if (props.style === 'block') {
+      return code.split('\n').map(line => `    ${line}`).join('\n') + '\n';
+    } else {
+      // Inline code must be at least 1 letter
+      if (code.length === 0) {
+        return '` `';
+      } else {
+        let backticks = '`'.repeat(getNumberOfRequiredBackticks(code));
+        return `${backticks}${code}${backticks}`;
+      }
+    }
+  }
+
+  *'html'(props: { type: string }, state: State): IterableIterator<string> {
+    state.push({ isPreformatted: true, htmlSafe: true });
+    let text = yield;
+    state.pop();
+    let html = text.join('');
+    if (props.type === 'block') {
+      return html + '\n';
+    }
+    return html;
   }
 
   /**
    * A list item is part of an ordered list or an unordered list.
    */
   *'list-item'(_, state: State): IterableIterator<string> {
-    let indent: string = '   '.repeat(state.get('indent'));
-    let rawItem: string[] = yield;
     let index: number = state.get('index');
-    let item: string = rawItem.join('').split('\n').map(line => indent + line).join('\n').trim();
-
-    if (state.get('type') === 'ordered-list') {
-      item = `${indent}${index}. ${item}`;
+    let delimiter = state.get('delimiter');
+    let marker: string = delimiter;
+    if (state.get('type') === 'numbered') {
+      marker = `${index}${delimiter}`;
       state.set('index', index + 1);
-    } else if (state.get('type') === 'unordered-list') {
-      item = `${indent}- ${item}`;
     }
+    let indent = ' '.repeat(marker.length + 1);
+    let text: string[] = yield;
+    let item: string = text.join('');
+    let firstCharacter = 0;
+    while (item[firstCharacter] === ' ') firstCharacter++;
+    let lines = item.split('\n');
+    lines.push(lines.pop().replace(/( )+$/, ''));
+    lines.unshift(lines.shift().replace(/^( )+/, ''));
+    let [first, ...rest] = lines;
 
-    return item;
+    item = ' '.repeat(firstCharacter) + first + '\n' + rest.map(line => indent + line).join('\n').replace(/( )+$/, '');
```

Comment:
trivial / pointless-micro-optimization - i'd probably replace the regex group w/ just a character set holding the space `[ ]`

Reply:
This only trims trailing spaces, would that do the same thing?

Reply:
yep - you have a capture group that you don't need, i think you could just use the space character but i feel like using the square braces for a character set makes it easier to spot with the eyes
- `/ +$/` should work but it's hard to realize what's going on
`/[ ]+$/` should be the same and is a tiny bit easier to recognize that it's a space character